### PR TITLE
Functions to merge a set of backup details

### PR DESCRIPTION
**Pull request by ashmrtn**
## Description

Go through the provided bases, load their backup details, and check if
any of the items in them need to be merged into the details for the
current backup.

Has a small amount of logic to treat moved items as updated.

Also include all the tests

Viewing by commit may be useful

## Does this PR need a docs update or release note?

- [ ] :white_check_mark: Yes, it's included
- [ ] :clock1: Yes, but in a later PR
- [x] :no_entry: No 

## Type of change

- [x] :sunflower: Feature
- [ ] :bug: Bugfix
- [ ] :world_map: Documentation
- [ ] :robot: Test
- [ ] :computer: CI/Deployment
- [ ] :hamster: Trivial/Minor

## Issue(s)

* #1800 

## Test Plan

- [ ] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: src/internal/operations/backup.go [hunk 1]
```diff
@@ -407,6 +397,119 @@ func consumeBackupDataCollections(
...
+				)
+			}
+
+			deets.Add(
+				newPath.String(),
+				newPath.ShortRef(),
+				newPath.ToBuilder().Dir().ShortRef(),
+				// TODO(ashmrtn): This may need updated if we start using this merge
+				// strategry for items that were cached in kopia.
+				newPath.String() != rr.String(),
```
Comment (ashmrtn) [inline src/internal/operations/backup.go:492]: @vkamra I'm not sure if moved items should count as updated. We can iterate on this

File: src/internal/operations/backup.go [hunk 2]
```diff
@@ -407,6 +397,119 @@ func consumeBackupDataCollections(
...
+
+			// Track how many entries we added so that we know if we got them all when
+			// we're done.
+			addedEntries++
+		}
+	}
+
+	if addedEntries != len(toMerge) {
+		return errors.Errorf(
+			"merging details: wanted %v entries but added %v",
```
Comment (ryanfkeepers) [inline src/internal/operations/backup.go:504]: For end user clarity
```suggestion
			"incomplete migration of backup details: found %v of %v expected items",
```

File: src/internal/operations/backup.go [hunk 3]
```diff
@@ -407,6 +397,119 @@ func consumeBackupDataCollections(
...
+
+	return false
+}
+
+func mergeDetails(
+	ctx context.Context,
+	ms *store.Wrapper,
+	detailsStore detailsReader,
+	mans []*kopia.ManifestEntry,
+	toMerge map[string]path.Path,
```
Comment (ryanfkeepers) [inline src/internal/operations/backup.go:417]: `toMerge` has, in this follow-up PR, revealed itself to be a bit ambiguous.  It describes the goal of the map, but not its contents.  For a dev who doesn't already grok the design at large, the intent is difficult to tie in to a mental model.  Something that more directly describes the value would make it more useful.

I honestly don't have a good replacement name off the top of my head.  `shortRefsFromPrevBackup`?

**Review by ryanfkeepers — APPROVED** [review]

File: src/internal/operations/backup.go [hunk 4]
```diff
@@ -407,6 +397,119 @@ func consumeBackupDataCollections(
 	return bu.BackupCollections(ctx, bases, cs, sel.PathService(), oc, tags)
 }
 
+func matchesReason(reasons []kopia.Reason, p path.Path) bool {
```
[the remark this answers is not in the document]
Comment (ashmrtn) [reply src/internal/operations/backup.go:400]: I was actually thinking we should factor our `Reason` from the kopia package, replace `OwnersCats` with it where needed, and do a little cleanup as to how kopia generates tags for a snapshot. I'm not sure if the `Reason <-> path` check should live in the path package though. That seems like it's adding more functions to an already fairly large interface

**Pull request merged**